frontend/src/components/widgets/DataFrame/EditingState.ts

```diff
@@ -0,0 +1,236 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { GridCell } from "@glideapps/glide-data-grid"
+
+import { notNullOrUndefined, isNullOrUndefined } from "src/lib/utils"
+
+import { BaseColumn } from "./columns"
+
+/**
+ * The editing state keeps track of all table edits applied by the user.
+ */
+class EditingState {
+  // row -> column -> GridCell
+  // Using [number, number] as a key for a Map would not work.
+  private editedCells: Map<number, Map<number, GridCell>> = new Map()
+
+  // List of rows represented by of column -> GridCell mappings
+  private addedRows: Array<Map<number, GridCell>> = []
+
+  // List of deleted row IDs
+  private deletedRows: number[] = []
+
+  // The original number of rows in the table (without potential additions & deletions)
+  private numRows = 0
+
+  constructor(numRows: number) {
+    this.numRows = numRows
+  }
+
+  /**
+   * Convert the current editing state to a JSON string.
+   *
+   * @param columns - The columns of the table
+   * @returns JSON string
+   */
+  toJson(columns: BaseColumn[]): string {
+    const columnsByIndex = new Map<number, BaseColumn>()
+    columns.forEach(column => {
+      columnsByIndex.set(column.indexNumber, column)
+    })
+
+    const currentState = {
+      edited_cells: {} as Record<string, any>,
+      added_rows: [] as Record<number, any>[],
+      deleted_rows: [] as number[],
+    }
+
+    // Prepare edited cells
+    this.editedCells.forEach(
+      (row: Map<number, GridCell>, rowIndex: number, _map) => {
+        row.forEach((cell: GridCell, colIndex: number, _map) => {
+          const column = columnsByIndex.get(colIndex)
+          if (column) {
+            currentState.edited_cells[`${rowIndex}:${colIndex}`] =
+              column.getCellValue(cell)
+          }
+        })
+      }
+    )
+
+    // Prepare added rows
+    this.addedRows.forEach((row: Map<number, GridCell>) => {
+      const addedRow: Record<number, any> = {}
+      row.forEach((cell: GridCell, colIndex: number, _map) => {
+        const column = columnsByIndex.get(colIndex)
+        if (column) {
+          const cellValue = column.getCellValue(cell)
+          if (notNullOrUndefined(cellValue)) {
+            addedRow[colIndex] = cellValue
+          }
+        }
+      })
+      currentState.added_rows.push(addedRow)
```

Comment:
Good catch :) Indeed, empty rows are not directly added. But it's not necessarily caused by this part of code. If all values are `null`/`missing` this would still add an empty mapping to the `added_rows`. But we currently don't trigger a rerun if a user just adds a new row ([trigger rerun in applyEdits is false](https://github.com/streamlit/streamlit/blob/7177652bc7728fb110f43990cae473f8a68a3ec3/frontend/src/components/widgets/DataFrame/hooks/useDataEditor.ts#LL115C5-L115C15)). Reruns are only triggered if cells get edited or rows get deleted, so it works like this:

https://user-images.githubusercontent.com/2852129/216393000-8d326824-88b4-4d9a-8a82-5b0db84f0828.mov

This behavior is something that we want to evaluate in the experimental release. 

> I was seeing it as adding rows for other people to fill out in FIXED mode

The app backend itself can add empty rows (by adding empty rows to the pandas dataframe) which also works with `FIXED` mode.